cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,331 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * ArtifactLocalizer is responsible for fetching, caching and unpacking artifacts requested by the worker pod. The HTTP
+ * endpoints are defined in {@link ArtifactLocalizerHttpHandlerInternal}. This class will run in the sidecar container
+ * that is defined by {@link ArtifactLocalizerTwillRunnable}.
+ *
+ * Artifacts will be cached using the following file structure:
+ *
+ * /PD_DIRECTORY/artifacts/<namespace>/<artifact-name>/<artifact-version>/<last-modified-timestamp>.jar
+ *
+ * Artifacts will be unpacked using the following file structure:
+ *
+ * /PD_DIRECTORY/unpacked/<namespace>/<artifact-name>/<artifact-version>/<last-modified-timestamp>/...
+ */
+public class ArtifactLocalizer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactLocalizer.class);
+  private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.RFC_1123_DATE_TIME;
+  // TODO: Move this directory name into a cConf property
+  private static final String PD_DIR = "data/";
+  private static final String LAST_MODIFIED_HEADER = HttpHeaderNames.LAST_MODIFIED.toLowerCase();
+
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+  private final String basePath;
+
+  @Inject
+  ArtifactLocalizer(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this(cConf, discoveryServiceClient, "");
+  }
+
+  @VisibleForTesting
+  public ArtifactLocalizer(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient, String basePath) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                         HttpRequestConfig.DEFAULT,
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+    this.basePath = basePath;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as well
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    Long lastModifiedTimestamp = getCurrentLastModifiedTimestamp(artifactId);
```

Comment:
It seems strange that the `getCurrentLastModifiedTimestamp` and the `fetchArtifact` is separated as two calls. From this method point of view, timestamp is not something it needs to care about.

Reply:
done, moved the call to `getCurrentLastModifiedTimestamp` into the `fetchArtifact` method and removed the cache delete method